core/blockchain.go

```diff
@@ -665,10 +665,11 @@ func SetReceiptsData(config *params.ChainConfig, block *types.Block, receipts ty
 		// The transaction hash can be retrieved from the transaction itself
 		receipts[j].TxHash = transactions[j].Hash()
 
-		tx, _ := transactions[j].AsMessage(signer)
 		// The contract address can be derived from the transaction itself
-		if MessageCreatesContract(tx) {
-			receipts[j].ContractAddress = crypto.CreateAddress(tx.From(), tx.Nonce())
+		if transactions[j].To() == nil {
+			// Deriving the signer is expensive, only do if it's actually needed
+			from, _ := types.Sender(signer, transactions[j])
```

Comment:
As mentioned privately (just for the reference), the transactions array is already matches to the block tx root, so unless the entire block is invalid, its's fine to ignore here. Also fast sync relies on PoW to verify the validity of headers not tx processing, so we don't sacrifice anything by trusting the data content of an accepted header.